src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/HtmlMarkupParser.cs

```diff
@@ -704,6 +704,26 @@ private bool TryRecoverStartTag(in SyntaxListBuilder<RazorSyntaxNode> builder, s
                     {
                         // There is no matching end void tag.
                         tagMode = MarkupTagMode.Void;
+
+                        // Try looking (until the end of the code block) for a matching end tag.
+                        // This needs to be supported in case the tag is actually a component.
+                        while (!EndOfFile && !At(SyntaxKind.RightBrace))
+                        {
+                            if (At(SyntaxKind.OpenAngle) && NextIs(SyntaxKind.ForwardSlash))
```

Comment:
I had realized this simple logic breaks for more complex cases like

```razor
@{
  <col>
  x = 1; // C# before, would be parsed as HTML now
  <script>const text = '</col>';</script>
}
```

I rewrote the fix, although it's more complicated now, so I'm not sure if it's worth the probably-niche use case (having a component named as one of the predefined HTML void tags).